### PR TITLE
cachable: make sure to clear caches between tests

### PR DESCRIPTION
**Pull request by apainintheneck**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew typecheck` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?

-----

We've been seeing some difficult to debug test failures recently related to not clearing some caches between tests. Normally we use the `Cachable` mixin to facilitate caching in classes and modules. I thought it'd be nice to automatically clear those caches between tests without having to create a list manually of all of the classes and modules that use this mixin.

This adds a registry for all modules and classes that cachable is included in. The registry allows us to programmatically clear all caches in between tests so that we don't forget to do that when adding a new class or refactoring code. The goal here is to reduce the number of flaky tests in the future. We only fill this registry when `HOMEBREW_TESTS` is set.

Marking draft to wait for discussion in https://github.com/Homebrew/brew/pull/16671 since this will depend on that.

### CODE REVIEW

File: Library/Homebrew/api/formula.rb [hunk 1]
```diff
@@ -10,9 +10,9 @@ module API
     #
     # @api private
     module Formula
-      class << self
-        include Cachable
+      extend Cachable
 
+      class << self
```
Comment (reitermarkus) [inline Library/Homebrew/api/formula.rb:15]: Might as well use `private_class_method` and move everything to `def self.*` here.

File: Library/Homebrew/extend/cachable.rb [hunk 2]
```diff
@@ -11,4 +11,43 @@ def cache
   def clear_cache
     cache.clear
   end
+
+  # Collect all classes that mix in Cachable so that those caches can be cleared in-between tests.
+  if ENV["HOMEBREW_TESTS"]
+    def self.included(klass)
+      raise ArgumentError, "Don't use Cachable with singleton classes" if klass.singleton_class?
```
Comment (reitermarkus) [inline Library/Homebrew/extend/cachable.rb:18]: Why does this need to be special-cased/forbidden?

File: Library/Homebrew/extend/cachable.rb [hunk 3]
```diff
@@ -11,4 +11,43 @@ def cache
...
+
+      super if defined?(super)
+    end
+
+    # Ignore classes that get inherited from a lot and that have
+    # caches that we don't need to clear on the class level.
+    IGNORE_INHERITED_CLASSES = %w[Formula Cask].freeze
```
Comment (reitermarkus) [inline Library/Homebrew/extend/cachable.rb:25]: Cask is never inherited.

File: Library/Homebrew/api/formula.rb (hunk 1 above)
Comment (reitermarkus) [reply Library/Homebrew/api/formula.rb:15]: Can be done in a follow-up PR though.

File: Library/Homebrew/extend/cachable.rb [hunk 4]
```diff
@@ -11,4 +11,43 @@ def cache
   def clear_cache
     cache.clear
   end
+
+  # Collect all classes that mix in Cachable so that those caches can be cleared in-between tests.
+  if ENV["HOMEBREW_TESTS"]
```
Comment (reitermarkus) [inline Library/Homebrew/extend/cachable.rb:16]: I wonder, could we implement this as a test helper with `allow(Cachable).to receive(:extended)`.

File: Library/Homebrew/api/formula.rb (hunk 1 above)
Comment (apainintheneck) [reply Library/Homebrew/api/formula.rb:15]: I didn't think that'd work but it seems fine. Good idea!

File: Library/Homebrew/extend/cachable.rb (hunk 2 above)
Comment (apainintheneck) [reply Library/Homebrew/extend/cachable.rb:18]: I don't think there is a method to get the original class from the singleton class. Forcing classes to extend this module instead of including it in the singleton class avoids that problem and is functionally equivalent.

File: Library/Homebrew/extend/cachable.rb (hunk 3 above)
Comment (apainintheneck) [reply Library/Homebrew/extend/cachable.rb:25]: Brain fart

File: Library/Homebrew/extend/cachable.rb (hunk 4 above)
Comment (apainintheneck) [reply Library/Homebrew/extend/cachable.rb:16]: I don't think so. It needs to be set up before other classes start mixing in this module. I thought of another way of doing this where we load a separate file in `test/support/extend/` with these cachable methods before loading `global.rb` and then we can be sure that this only gets loaded in the test environment.

Comment (apainintheneck) [reply Library/Homebrew/extend/cachable.rb]: See https://github.com/Homebrew/brew/pull/16746/commits/909828caa1565da181aa877613c5992efa860c99

**Comment by apainintheneck** [comment]
I'm marking this as ready for review since it no longer changes any of the same files as https://github.com/Homebrew/brew/pull/16671 and should be useful whatever we decide to do there.

File: Library/Homebrew/api/formula.rb (hunk 1 above)
Comment (apainintheneck) [reply Library/Homebrew/api/formula.rb:15]: Sorry, I got confused. Yeah, I'll update that too.

File: Library/Homebrew/extend/cachable.rb [hunk 5]
```diff
@@ -1,4 +1,4 @@
-# typed: strict
+# typed: true
```
Comment (apainintheneck) [inline Library/Homebrew/extend/cachable.rb:1]: TODO: Revert this change since it's no longer necessary.

File: Library/Homebrew/api/formula.rb (hunk 1 above)
Comment (apainintheneck) [reply Library/Homebrew/api/formula.rb:15]: Might not be worth killing the git blame though. I'll do it anyway and see if anyone cares.

Comment (apainintheneck) [reply Library/Homebrew/api/formula.rb]: Addressed in https://github.com/Homebrew/brew/pull/16746/commits/bea2dc65fec2ef41c90918a541fef6deb3e15142

**Review by MikeMcQuaid — APPROVED** [review]
Good idea, thanks again @apainintheneck!

File: Library/Homebrew/test/spec_helper.rb [hunk 6]
```diff
@@ -202,16 +204,7 @@
   config.around do |example|
     Homebrew.raise_deprecation_exceptions = true
 
-    Formulary.clear_cache
-    Tap.each(&:clear_cache)
```
Comment (reitermarkus) [inline Library/Homebrew/test/spec_helper.rb:206]: @apainintheneck, this needs to stay.

Comment (apainintheneck) [reply Library/Homebrew/test/spec_helper.rb:206]: Is it really necessary to clear taps at the instance level if we're already clearing `Tap`, `CoreTap` and `CoreCaskTap` at the class level? I guess `.clear_cache` doesn't clear the core tap instances, right?

Comment (apainintheneck) [reply Library/Homebrew/test/spec_helper.rb:206]: That is the `@instance` instance variable.

Comment (apainintheneck) [reply Library/Homebrew/test/spec_helper.rb:206]: https://github.com/Homebrew/brew/pull/16760
